fpga_interchange/fasm_generators/generic.py

```diff
@@ -0,0 +1,133 @@
+#
+# SPDX-License-Identifier: ISC
+from collections import namedtuple
+
+from fpga_interchange.route_stitching import flatten_segments
+from fpga_interchange.physical_netlist import PhysicalPip
+
+PhysCellInstance = namedtuple(
+    'CellInstance', 'cell_type site_name tile_name sites_in_tile attributes')
+
+
+class FasmGenerator():
+    def __init__(self, interchange, device_resources, log_netlist_file,
+                 phy_netlist_file):
+        with open(device_resources, "rb") as f:
+            self.device_resources = interchange.read_device_resources(f)
+
+        with open(log_netlist_file, "rb") as f:
+            self.logical_netlist = interchange.read_logical_netlist(f)
+
+        with open(phy_netlist_file, "rb") as f:
+            self.physical_netlist = interchange.read_physical_netlist(f)
+
+        self.physical_cells_instances = dict()
+        self.logical_cells_instances = dict()
+
+        self.routing_pips_features = list()
+        self.cells_features = list()
+
+    def build_log_cells_instances(self):
+        """
+        Fills a dictionary with the logical cells with their attribute map
+        """
+
+        lib = self.logical_netlist.libraries["work"]
+        for cell in lib.cells.values():
+            for cell_instance, cell_obj in cell.cell_instances.items():
+
+                cell_name = cell_obj.cell_name
+                cell_attrs = cell_obj.property_map
+
+                assert cell_instance not in self.logical_cells_instances, (
+                    cell_instance, self.logical_cells_instances.keys())
+                self.logical_cells_instances[cell_instance] = cell_attrs
+
+    def build_phys_cells_instances(self):
+        """
+        Fills a dictionary for handy lookups of the placed sites and the corresponding tiles.
+
+        The dictionary contains PhysCellInstance objects which are filled with the attribute
+        map from the logical netlist
+        """
+
+        for placement in self.physical_netlist.placements:
+            cell_name = placement.cell_name
+            cell_type = placement.cell_type
+            site_name = placement.site
+            tile_name = self.device_resources.get_tile_name_at_site_name(
+                site_name)
+            tile = self.device_resources.tile_name_to_tile[tile_name]
+            sites_in_tile = tile.site_names
+
+            cell_attr = self.logical_cells_instances.get(cell_name, None)
+
+            self.physical_cells_instances[cell_name] = PhysCellInstance(
+                cell_type=cell_type,
+                site_name=site_name,
+                tile_name=tile_name,
+                sites_in_tile=sites_in_tile,
+                attributes=cell_attr)
+
+    def fill_pip_features(self, site_thru_features=dict()):
+        """
+        This function generates all features corresponding to the physical routing
+        PIPs present in the physical netlist.
+
+        At the moment, the convention for the PIPs FASM features is:
```

Comment:
To consider:
 - prjoxide (and prjuray) currently uses `<TILE_NAME>.PIP.<WIRE1>.<WIRE0>` to distinguish pips better from other features
 - prjoxide bitstream tile names don't correspond to the interchange tile names, because prjoxide can have multiple bitstream tiles at one grid location that get merged into one interchange tile

I think some extra annotations in the database could solve this, for example a per-pip fasm prefix (`StringIdx` means this will be cheap to store as many pips will share the same prefix).

Reply:
Right, for the additional `PIP` annotation in the feature there might be two solutions:
- pass a pattern to correctly fill the FASM features, as all PIPs should have a tile_prefix and two wires, this should be easily doable
- begin a transition process to have a standard in the various databases, such that all adopt the same convention.

Reply:
I have added a way to specify the FASM feature formatting so to add the `.PIP.` part.

I think that enabling the FASM prefixes from the device resources database can be done in follow-up PRs